src/pcp/slabinfo/pcp-slabinfo.1
'\" t
.\"
.\" Man page for pcp-slabinfo
.\" Copyright (c) 2023 Oracle and/or its affiliates.
.\" DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.TH PCP-SLABINFO 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pcp-slabinfo\f1 \- Report statistics for kernel slab allocator
.SH SYNOPSIS
\f3pcp\f1
[\f2pcp\ options\f1]
\f3slabinfo\f1
[\f3\-a\f1 \f2archive\f1]
[\f3\-s\f1 \f2samples\f1]
[\f3\-Z\f1 \f2timezone\f1]
.SH DESCRIPTION
The
.B pcp-slabinfo
command is used for viewing different stats related to slab.
It helps users analyze useful information related to the slab allocator.
The information includes the total number of objects that are currently
active, allocated objects, pages per slab, etc.
By default,
.B pcp-slabinfo
reports live data for the local host.
.PP
The statistics shown are as follows:
.TS
lfB  lfB
l    lx.
HEADER	DESCRIPTION
_	_
active_objs	T{
.ad l
The number of objects that are currently active (i.e., in use)
T}
num_objs	T{
.ad l
The total number of allocated objects (i.e. objects that are both in use and not in use).
T}
objsize	The size of objects in this slab, in bytes.
objperslab	The number of objects stored in each slab.
pagesperslab	The number of pages allocated for each slab
active_slabs	The number of active slabs.
num_slabs	The total number of slabs.
.TE
.SH OPTIONS
.TP
\fB\-a\fR \fIfile\fR, \fB\-\-archive\fR=\fIfile\fR
Fetch /proc/slabinfo for a specified archive \fIfile\fR
.TP
\fB\-s\fR \fIcount\fR, \fB\-\-samples\fR=\fIcount\fR
Get the slabinfo statistics for specified number of samples \fIcount\fR
.TP
\fB\-V\fR, \fB\-\-version\fR
Display version number and exit.
.TP
\fB\-z\fR, \fB\-\-hostzone\fR
Use the local timezone of the host that is the source of the
performance metrics, as identified by either the
.B \-h
or the
.B \-a
options.
The default is to use the timezone of the local host.
.TP
\fB\-Z\fR \fItimezone\fR, \fB\-\-timezone\fR=\fItimezone\fR
Use
.I timezone
for the date and time.
.I Timezone
is in the format of the environment variable
.B TZ
as described in
.BR environ (7).
.TP
\fB\-?\fR, \fB\-\-help\fR
Display usage message and exit.
.SH NOTES
.B pcp-slabinfo
collects information from
.IR /proc/slabinfo .
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fP are used to parameterize the file and directory names used by PCP. On each installation, the file \fI/etc/pcp.conf\fP contains the local values for these variables. The \fB$PCP_CONF\fP variable may be used to specify an alternative configuration file, as described in \fBpcp.conf\fP(5).
.PP
For environment variables affecting PCP tools, see \fBpmGetOptions\fP(3).
.SH SEE ALSO
.BR PCPIntro (1),
.BR pcp (1),
.BR pmParseInterval (3),
.BR environ (7).

.\" control lines for scripts/man-spell
.\" +ok+ pagesperslab active_slabs active_objs objperslab num_slabs
.\" +ok+ slabinfo {from pcp-slabinfo}
.\" +ok+ num_objs objsize


CHANGELOG
pcp-7.0.0 (1 July 2025)
    - Work-in-progress.

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.2 (6 November 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdabpf: avoid a too-low rlimit setting within pmdabpf
	pmdaproc: add an access control mechanism for remote sampling
	pmdalinux: ensure divide-by-zero cannot happen with tmpfs.full
	pmdalinux: additional MPTcpExt netstat metrics
	pmdastatsd: keep correct track of pmns mapped metric count
	pmdaopenmetrics: run as pcp user rather than root by default
	pmdaopenmetrics: ensure correct cache file permissions used
    - Server-side utilities and log management scripts:
	libpcp, pmlogger: add support for zstd(1) compression of archives
	pmlogbasename: new, maps filename to PCP archive basename
	pmlogcompress: new, archive compression/decompression utility
	pmlogcheck: performance improvements
	pmlogcheck:  allow multiple archive options on command line
	pmlogextract: extend -v option with the full pmlogger syntax
	pmlogextract: fix for reporting file size on 32-bit platforms
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix 64-bit big endian v3 archive timestamp decoding
	libpcp: change in archive value interpolation for non-counter metrics
	python api: allow timeval and timespec init with float/datetime
    - Client tools and utilities:
	pminfo: minor error path reworking
	pcp-buddyinfo: addition of a --no-interpol option
	pcp-ps: added option to view full task name with its argument
    - Build, infrastructure and packaging updates:
	build: deal with botched deb package names from Debian
	build: ensure pmie config.default created with correct ownership
	systemd: move pcp-reboot-init rpm spec logic from zeroconf to pcp
	selinux: ensure farm pmlogger/pmie processes labelled correctly
    - Documentation:
	docs: refine pmdaproc -A description details
	docs: make pmstat(1) match the code for -t defaults

pcp-6.3.1 (17 September 2024)
    - Client tools and utilities:
	pcp2arrow: fix extraction of values for metrics with indoms
	pcp-htop: fix platform setting enabling screen tabs
	pcp-htop: update to latest upstream bug fixes
	pcp-meminfo: added timestamp string for consistency
    - PMDA additions, enhancements and bug fixes:
	pmdaopenmetrics: add vLLM metrics support by default
	pmdahacluster: handle pacemaker 2.1.6 crm_mon output changes
	pmdahacluster: handle new style Pacemaker role descriptions
	pmdalinux: MPJoinSynBackupRx, MPJoinSynAckBackupRx netstat metrics
	pmdalinux: sysfs fallback from hinv.cpu.clock metric
	pmdaproc: lazy initialisation for accounting metrics
	pmdapmcd: expose pmcd limits as new metrics
	pmdammv: use isnan instead of open-coding NaN checks
    - Server-side utilities and log management scripts:
	pmcd: put an explicit cap on max number of fetched PMIDs
	pmcd: put an explicit cap on max number of contexts per client
	pmcd: cleanup caching of local hostname string
	pmpost: guard against possible symlink attack
	pmproxy: transition to dual Redis/Valkey support
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp_web: fix a race in webgroup context teardown on timeout
	libpcp_web: transition to dual Redis/Valkey support
	libpcp: auto-load the proc PMDA in local context mode
	libpcp: fix buffer overflow in __pmServerAddInterface
	libpcp: refactor the attr PDU content checking
	libpcp: desc PDU consistency check and error message rework
	libpcp: text PDU consistency check and error message rework
	libpcp: creds PDU consistency check and error message rework
	libpcp: label PDU consistency check and error message rework
	libpcp: instance PDU consistency check and error message rework
	libpcp: rework some fetch PDU integrity checks and error messages
	libpcp: reworked __pmDecodeIDList integrity checks and warnings
	libpcp: ensure no fd leakage to child from __pmProcessPipe
	libpcp: ensure container names contain limited characters
	libpcp: fix buffer overrun test in __pmDecodeValueSet
	libpcp: hardening of the results PDU handling
	libpcp: hardening of the creds PDU handling
	libpcp: add new __pmCtlDebug() function
	python API: fix MMV metric text, indom and semantics handling
	python API: add log import timestamp convenience mechanisms
    - Build, infrastructure and packaging updates:
	build: tweak Makepkgs as dpkg-parsechangelog(1) is a bit picky
	build: drop gfs2 PMDA from build if distro vendor opts-out
	build: update denki.conf packaging location for consistency
    - Documentation:
	man pages: tidy up references to $PCP_DEBUG
	man pages: __pmHash* added to man3i
	man pages: pmnewcontext(3) unix: and local: option details
	docs: man page updates transitioning away from Redis
	docs: correct pmlogger(1) man page reference to section 5 page
	docs: update the books to refer to key servers more generically
	pmrep et al man pages: describe derived metric troubleshooting

pcp-6.3.0 (30 July 2024)
    - Client tools and utilities:
	pcp-htop: add AMD GPU metrics support
	pcp-htop: fix Available Meter column list in Setup/F2
	pcp-htop: fix nested derived metric definitions for devices
	pcp-htop: fix label/color botch in the MySQL keys Meter config
	pcp-xsos: fast system overview for Red Hat customer support
	pcp-atop: fix units conversion for ipc.shm metric values
	pcp2openmetrics: stricter adherence to the specification
	pmview: developer preview only, incl. mpvis, dkvis, osvis
	pmstat: include guest time in extra CPU stats
	pmrep conf: add support for pmstat -x alike metricset
	pmrep conf: fix collectl cpu metrics to match current versions
	pmrep conf: add guest cpu time to vmstat output where missing
	pmrep conf: add metricset to mimic pmstat(1)
	pmrep conf: fix cpu metrics in sar/vmstat metricset formulas
	pmconfig: add -L y2038_safe to know if 8-byte time_t used
    - PMDA additions, enhancements and bug fixes:
	pmdaamdgpu: metrics from libdrm and libdrm-amdgpu libraries

src/pcp/slabinfo/pcp-slabinfo.py
#!/usr/bin/env pmpython
#
# Copyright (c) 2023 Oracle and/or its affiliates.
# DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
# pylint: disable=bad-whitespace,too-many-lines,bad-continuation
# pylint: disable=too-many-arguments,too-many-positional-arguments
# pylint: disable=redefined-outer-name,unnecessary-lambda
#
import signal
import sys
import time
from pcp import pmapi, pmcc
from cpmapi import PM_CONTEXT_ARCHIVE

SYS_MECTRICS= ["kernel.uname.sysname","kernel.uname.release",
               "kernel.uname.nodename","kernel.uname.machine","hinv.ncpu"]
SLABSTAT_METRICS = ["mem.slabinfo.slabs.active",
                    "mem.slabinfo.slabs.pages_per_slab","mem.slabinfo.slabs.total_size",
                    "mem.slabinfo.slabs.objects_per_slab","mem.slabinfo.objects.active",
                    "mem.slabinfo.objects.total","mem.slabinfo.objects.size",
                    "mem.slabinfo.slabs.total" ]
ALL_METRICS = SLABSTAT_METRICS + SYS_MECTRICS
def adjust_length(name):
    return name.ljust(25)
class ReportingMetricRepository:

    def __init__(self,group):
        self.group=group
        self.current_cached_values = {}

    def __sorted(self,data):
        return dict(sorted(data.items(), key=lambda item: item[0].lower()))

    def __fetch_current_value(self,metric):
        val=dict(map(lambda x: (x[1], x[2]), self.group[metric].netValues))
        val=self.__sorted(val)
        return dict(val)

    def current_value(self,metric):
        if not metric in self.group:
            return None
        if self.current_cached_values.get(metric) is None:
            first_value=self.__fetch_current_value(metric)
            self.current_cached_values[metric]=first_value
        return self.current_cached_values[metric]

class SlabStatUtil:
    def __init__(self,metrics_repository):
        self.__metric_repository=metrics_repository
        self.report=ReportingMetricRepository(self.__metric_repository)

    def active_objs(self):
        return self.report.current_value('mem.slabinfo.objects.active')

    def objs_size(self):
        return self.report.current_value('mem.slabinfo.objects.size')

    def obj_total(self):
        return self.report.current_value('mem.slabinfo.objects.total')

    def slab_total_size(self):
        return self.report.current_value('mem.slabinfo.slabs.total_size')

    def objects_per_slab(self):
        return self.report.current_value('mem.slabinfo.slabs.objects_per_slab')

    def page_per_slab(self):
        return self.report.current_value('mem.slabinfo.slabs.pages_per_slab')

    def slab_active(self):
        return self.report.current_value('mem.slabinfo.slabs.active')

    def slabs_total(self):
        return self.report.current_value('mem.slabinfo.slabs.total')

    def names(self):
        data = self.report.current_value('mem.slabinfo.objects.active')
        return data.keys()

class SlabinfoReport(pmcc.MetricGroupPrinter):
    def __init__(self,opts,group):
        self.opts = opts
        self.group = group
        self.samples = opts.samples
        self.context = opts.context

    def __get_ncpu(self, group):
        return group['hinv.ncpu'].netValues[0][2]

    def __print_machine_info(self, context):
        timestamp = self.group.pmLocaltime(context.timestamp.tv_sec)
        # Please check strftime(3) for different formatting options.
        # Also check TZ and LC_TIME environment variables for more
        # information on how to override the default formatting of
        # the date display in the header
        time_string = time.strftime("%x", timestamp.struct_time())
        header_string = ''
        header_string += context['kernel.uname.sysname'].netValues[0][2] + '  '
        header_string += context['kernel.uname.release'].netValues[0][2] + '  '
        header_string += '(' + context['kernel.uname.nodename'].netValues[0][2] + ')  '
        header_string += time_string + '  '
        header_string += context['kernel.uname.machine'].netValues[0][2] + '  '
        print("%s  (%s CPU)" % (header_string, self.__get_ncpu(context)))

    def __print_header(self,header_indentation,value_indentation):
        print("TimeStamp"+ " "+header_indentation + "Name" + " "*22 +value_indentation+ "active_objs" +
              value_indentation +"num_objs" + value_indentation + "objsize byte" + value_indentation +
              "objperslab" + value_indentation +"pagesperslab" + value_indentation + "active_slabs" +
              value_indentation + "num_slabs")

    def __print_values(self,timestamp,header_indentation,\
                     value_indentation,slabstatus):
        names=slabstatus.names()
        active_objects=slabstatus.active_objs()
        num_objs=slabstatus.objs_size()
        objsize=slabstatus.obj_total()
        objperslab=slabstatus.objects_per_slab()
        pageperslab=slabstatus.page_per_slab()
        active_slabs=slabstatus.slab_active()
        num_slabs=slabstatus.slabs_total()
        for name in names:
            data = adjust_length(name) if len(name) < 25 else name
            print("%s %s %s %s %8d %s %8d %s %8d %s %7d %s %9d %s %8d %s %8d"
                                    % (timestamp, header_indentation,data, value_indentation, active_objects[name],
                                       value_indentation,objsize[name],value_indentation, num_objs[name],
                                       value_indentation, objperslab[name],value_indentation,pageperslab[name],
                                       value_indentation, active_slabs[name], value_indentation, num_slabs[name]))

    def print_report(self,group,timestamp,header_indentation,value_indentation,manager_slabinfo):
        def __print_slab_status():
            slabstatus = SlabStatUtil(manager_slabinfo)
            if slabstatus.names():
                try:
                    self.__print_machine_info(group)
                    self.__print_header(header_indentation, value_indentation)
                    self.__print_values(timestamp, header_indentation, value_indentation, slabstatus)
                except IndexError:
                    print("Incorrect machine info due to some missing metrics")
                return
            else:
                pass

        if self.context != PM_CONTEXT_ARCHIVE and self.samples is None:
            __print_slab_status()
            sys.exit(0)
        elif self.context == PM_CONTEXT_ARCHIVE and self.samples is None:
            __print_slab_status()
        elif self.samples >=1:
            __print_slab_status()
            self.samples-=1
        else:
            pass

    def report(self, manager):
        group = manager["sysinfo"]
        self.samples = self.opts.pmGetOptionSamples()
        t_s = group.contextCache.pmLocaltime(int(group.timestamp))
        timestamp = time.strftime(SlabinfoOptions.timefmt, t_s.struct_time())
        header_indentation = "        " if len(timestamp) < 9 else (len(timestamp) - 7) * " "
        value_indentation = ((len(header_indentation) + 9) - len(timestamp)) * " "
        self.print_report(group,timestamp,header_indentation,value_indentation,manager['slabinfo'])

class SlabinfoOptions(pmapi.pmOptions):
    timefmt = "%H:%M:%S"
    def __init__(self):
        pmapi.pmOptions.__init__(self, "a:s:Z:zV?")
        self.pmSetLongOptionHeader("General options")
        self.pmSetLongOptionHostZone()
        self.pmSetLongOptionTimeZone()
        self.pmSetLongOptionHelp()
        self.pmSetLongOptionSamples()
        self.pmSetLongOptionVersion()
        self.samples=None
        self.context=None

if __name__ == '__main__':
    try:
        opts = SlabinfoOptions()
        mngr = pmcc.MetricGroupManager.builder(opts,sys.argv)
        opts.context=mngr.type
        missing = mngr.checkMissingMetrics(ALL_METRICS)
        if missing is not None:
            sys.stderr.write('Error: not all required metrics are available\nMissing %s\n' % missing)
            sys.exit(1)
        mngr["slabinfo"] = SLABSTAT_METRICS
        mngr["sysinfo"] = SYS_MECTRICS
        mngr["allinfo"]=ALL_METRICS
        mngr.printer = SlabinfoReport(opts,mngr)
        sts = mngr.run()
        sys.exit(sts)
    except pmapi.pmErr as error:
        sys.stderr.write('%s\n' % (error.message()))
    except pmapi.pmUsageErr as usage:
        usage.message()
        sys.exit(1)
    except IOError:
        signal.signal(signal.SIGPIPE, signal.SIG_DFL)
    except KeyboardInterrupt:
        pass
